zng/docs/zeek-compat.md

```diff
@@ -1,138 +1,188 @@
+#empty_field	(empty)
+#unset_field	-
+#fields	my_bool	my_count	my_int	my_double	my_time	my_interval	my_printable_string	my_bytes_string	my_port	my_addr	my_subnet	my_enum	my_set	my_vector	my_record.name	my_record.age
+#types	bool	count	int	double	time	interval	string	string	port	addr	subnet	enum	set[string]	vector[string]	string	count
+T	123	456	123.4560	1592502151.123456	123.456	smile\xf0\x9f\x98\x81smile	\x09\x07\x04	80	127.0.0.1	10.0.0.0/8	tcp	things,in,a,set	order,is,important	Jeanne	122
+
+$ zq -t zeek_types.log 
+#zenum=string
+#0:record[my_bool:bool,my_count:uint64,my_int:int64,my_double:float64,my_time:time,my_interval:duration,my_printable_string:bstring,my_bytes_string:bstring,my_port:port,my_addr:ip,my_subnet:net,my_enum:zenum,my_set:set[bstring],my_vector:array[bstring],my_record:record[name:bstring,age:uint64]]
+0:[T;123;456;123.456;1592502151.123456;123.456;smile😁smile;\x09\x07\x04;80;127.0.0.1;10.0.0.0/8;tcp;[a;in;set;things;][order;is;important;][Jeanne;122;]]
+
+$ zq -t zeek_types.log | zq -f zeek -
+#separator \x09
+#set_separator	,
+#empty_field	(empty)
+#unset_field	-
+#fields	my_bool	my_count	my_int	my_double	my_time	my_interval	my_printable_string	my_bytes_string	my_port	my_addr	my_subnet	my_enum	my_set	my_vector	my_record.name	my_record.age
+#types	bool	count	int	double	time	interval	string	string	port	addr	subnet	enum	set[string]	vector[string]	string	count
+T	123	456	123.456	1592502151.123456	123.456	smile\xf0\x9f\x98\x81smile	\x09\x07\x04	80	127.0.0.1	10.0.0.0/8	tcp	a,in,set,things	order,is,important	Jeanne	122
 ```
-where `<separator>` is intialized to a space character at the beginning of the
-stream, then overridden by the `#separator` directive.
-
-In the legacy format, the `#separator` character and the `#set_separator` character
-define how to parse both a legacy value line and a legacy descriptor.
-
-Every legacy value line corresponds to a `record` type defined by the
-fields and types directives possibly modified for the `#path` directive
-as described below.
-
-Record types may not be used in the `#types` directive,
-which means there is no need to recursively parse the `set` and `vector`
-container values (`set` and `vector` values are split according to
-the `#set_separator` character).
-
-
-
-## Legacy Values
-
-If a value line begins with `#`, then this
-character must be escaped.
-(Such lines can only exist as legacy values and not regular values
-as regular values always begin with an integer descriptor.)
 
+## Type-Specific Details
+
+As `zq` acts as a reference implementation for ZNG, it's helpful to understand
+the following subtleties of how it reads the following Zeek data types into
+ZNG equivalents and writes them back out again in Zeek log format. Other
+ZNG implementations (should they exist) may handle these differently.
+
+### `double`
+
+As they do not affect accuracy, "trailing zero" decimal digits on Zeek `double`
+values will _not_ be preserved (e.g. `123.4560` becomes `123.456`).
```

Comment:
```suggestion
values will _not_ be preserved when they are formatted into a string (e.g. `123.4560` becomes `123.456`).
```

this isn't really about doubles but about their string representation, so i'm not entirely sure it belongs next to the other points below.

Reply:
@henridf:  I'm inclined to take your suggestion and leave it rather than dropping the section as you seem to be implying you might prefer. Due to my obsessive nature to try and please everybody, lemme try out my logic on you anyway about why I think it's still appropriate. 😄 

1. I agree that it's about string representation. However, as Zeek TSV effectively _is_ string representation, and this doc is about compatibility with that, it seems like it becomes appropriate to address. With other generic UNIX pipeline tools or domain-specific ones like `zeek-cut`, it's usually reasonable for a user to assume that the data they pass through some in/out pipeline is going to be undisturbed unless they invoke some option that's explicitly intended to change the data. Thinking along these lines, I took the role here of a user assuming "I should be able to send my Zeek data into `zq`, no-op it, and get the same Zeek data back, byte-for-byte". If this is seen as something valid to expect, then I'd expect that for any areas where we _don't_ meet that assumption, we should be able to defend our behaviors as beneficial or harmless. That's what I'm trying to do here for both the "cosmetic" cases I cite here for `double` and `set`: We feel it's mathematically defensible what we're doing from an _accuracy_ perspective, so, as a user, they should just expect it. If someone was super offended by it, they could raise it as a concern or pursue their own ZNG implementation that behaves differently.

2. In terms of mathematical arguments, I have flashbacks to high school chemistry class and having the concept of "significant digits" pounded into our heads. That is, we were taught that there truly _is_ a difference between `123.4560` and `123.456` because the former clearly communicates a measurement was taken out to the 4th decimal place and the it happened to fall perfectly on a 0, as opposed to the latter which _might_ have been the same measurement or it might have been subject to rounding to the 3rd decimal place. An obsessive person might defend Zeek's extra zeros on these grounds and say that, while we're free to represent them internally however is necessary to maintain mathematical accuracy, we'd have a responsibility to maintain record of the extra zeros in the event that we're asked to print it back out again in string representation undisturbed. And, of course, I feel we can defensibly counter that the compute and complexity to achieve that would not be worth the effort, especially when we also consider that binary representations can save storage.

Rather than get into all that, I felt like it's worthwhile to just tell the user to expect it so that way they don't waste time digging into it if they ever do a `diff` before/after, and assume nobody will ever complain.